File: mechasnippet/matcher.py
"""Filtro de snippets en tiempo real.

Búsqueda insensible a mayúsculas y tildes. Cada palabra del query debe
aparecer (en el nombre o el contenido). El ranking prioriza coincidencias en
el nombre por sobre el contenido, para que lo más relevante quede arriba.
"""

import unicodedata


def _norm(text):
    """Minúsculas y sin tildes, para comparar de forma robusta."""
    text = unicodedata.normalize("NFKD", text)
    text = "".join(ch for ch in text if not unicodedata.combining(ch))
    return text.lower()


def filter_snippets(items, query):
    """Devuelve la lista de (nombre, contenido) que coincide con `query`, ordenada.

    Sin query, devuelve todos los items en su orden original.
    """
    q = _norm(query).strip()
    if not q:
        return list(items)

    tokens = q.split()
    scored = []
    for name, content in items:
        norm_name = _norm(name)
        norm_content = _norm(content)
        haystack = norm_name + "\n" + norm_content

        if not all(token in haystack for token in tokens):
            continue

        # Menor score = más relevante.
        if norm_name.startswith(q):
            score = 0
        elif q in norm_name:
            score = 1
        elif all(token in norm_name for token in tokens):
            score = 2
        else:
            score = 3

        scored.append((score, norm_name, (name, content)))

    scored.sort(key=lambda row: (row[0], row[1]))
    return [row[2] for row in scored]

File: Sources/MechaSnippetCore/Matcher.swift
import Foundation

public struct Matcher {
    public static func normalize(_ text: String) -> String {
        text.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: nil)
    }

    public static func filter(
        _ snippets: [Snippet],
        query: String,
        normalizedCache: [UUID: (name: String, content: String)]? = nil
    ) -> [Snippet] {
        let q = normalize(query).trimmingCharacters(in: .whitespaces)
        if q.isEmpty { return snippets }
        let tokens = q.split(separator: " ").map(String.init)

        var scored: [(score: Int, name: String, snippet: Snippet)] = []
        for s in snippets {
            let normName = normalizedCache?[s.id]?.name ?? normalize(s.name)
            let normContent = normalizedCache?[s.id]?.content ?? normalize(s.content)
            let haystack = normName + "\n" + normContent
            guard tokens.allSatisfy({ haystack.contains($0) }) else { continue }

            let score: Int
            if normName.hasPrefix(q) { score = 0 }
            else if normName.contains(q) { score = 1 }
            else if tokens.allSatisfy({ normName.contains($0) }) { score = 2 }
            else { score = 3 }
            scored.append((score, normName, s))
        }
        scored.sort { $0.score != $1.score ? $0.score < $1.score : $0.name < $1.name }
        return scored.map(\.snippet)
    }
}
